rasa/core/validator.py

```diff
@@ -0,0 +1,178 @@
+from rasa.core.domain import Domain
+from rasa.nlu.training_data import load_data, TrainingData
+from rasa.core.training.dsl import StoryFileReader, StoryStep
+from rasa.core.training.dsl import UserUttered
+from rasa.core.training.dsl import ActionExecuted
+
+logger = logging.getLogger(__name__)
+
+
+class Validator(object):
+    """A class used to verify usage of intents and utterances."""
+
+    def __init__(self, domain: Domain, intents: TrainingData, stories: List[StoryStep]):
+        """Initializes the Validator object. """
+
+        self.domain = domain
+        self.intents = intents
+        self.valid_intents = []
+        self.valid_utterances = []
+        self.stories = stories
+
+    @classmethod
+    def from_files(
+        cls, domain_file: Text, nlu_data: Text, story_data: Text
+    ) -> "Validator":
+        """Create an instance from the domain, nlu and story files."""
+
+        domain = Domain.load(domain_file)
+        loop = asyncio.new_event_loop()
+        stories = loop.run_until_complete(
+            StoryFileReader.read_from_folder(story_data, domain)
+        )
+        intents = load_data(nlu_data)
+        return cls(domain, intents, stories)
+
+    def _search(self, vector: List[Any], searched_value: Any) -> bool:
+        """Searches for a element in a vector."""
+
+        vector.append(searched_value)
+        count = 0
+        while searched_value != vector[count]:
+            count += 1
+        if count == len(vector) - 1:
+            return False
+        else:
+            return True
+
+    def verify_intents(self):
+        """Compares list of intents in domain with intents in NLU training data."""
+
+        domain_intents = []
+        nlu_data_intents = []
+
+        for intent in self.domain.intent_properties:
+            domain_intents.append(intent)
+
+        for intent in self.intents.intent_examples:
+            nlu_data_intents.append(intent.data["intent"])
+
+        for intent in domain_intents:
+            found = self._search(nlu_data_intents, intent)
+            if not found:
+                logger.error(
+                    "The intent '{}' is listed in the domain file, but "
+                    "is not found in the NLU training data.".format(intent)
+                )
+            else:
+                self.valid_intents.append(intent)
+
+        for intent in nlu_data_intents:
+            found = self._search(domain_intents, intent)
+            if not found:
+                logger.error(
+                    "The intent '{}' is in the NLU training data, but "
+                    "is not listed in the domain.".format(intent)
+                )
+
+    def verify_intents_in_stories(self):
+        """Checks intents used in stories.
+
+        Verifies if the intents used in the stories are valid, and whether
+        all valid intents are used in the stories."""
+
+        if self.valid_intents == []:
+            self.verify_intents()
+
+        stories_intents = []
```

Comment:
let's use sets for these. This applies to domain_intents, files_intents, stories_intents, and probably all of the utterance lists as well. The idea is that sets won't add the same intents to the list every time it sees one, just the first time. This should speed up the second round when we then iterate over this list in (in my testing env with the demobot, the list of stories intents is of length 5530, while the set is of length 65).

You can initialize a set with
stories_intents = set()
Add to it via
stories_intents.add(item)
and iterate over them just like lists.

This should speed up and clean up duplicates when iterating over these sets.